File: python/src/imc_opcon.py
"""Control of main numerical calculation."""

import pickle
# import matplotlib.pyplot as plt

import imc_update
import imc_source
import imc_tally
import imc_track

import imc_global_io_data as io
import imc_global_mesh_data as mesh
import imc_global_phys_data as phys
import imc_global_time_data as time


def run():
    """
    Control calculation for fcimc.

    Timestep loop is within this function.
    """
    # Set plot times
    plottimes = [6.0, 15.0, 90.0]
    plottimenext = 0

    # Open output file
    fname = open(io.output_file, "wb")

    # Set an initial temperature field
    mesh.temp[:] = mesh.temp0[:]  # keV

    # Loop over timesteps

    time.time = 0.0

    for time.step in range(0, time.ns):

        print("\n" + "=" * 79)
        print("Timestep  {:4d}".format(time.step))
        print("Time (s)  {:24.16E}".format(time.time))
        print("Time (cm) {:24.16E}".format(time.time * phys.c))
        print("=" * 79)

        # Update temperature-dependent quantities
        imc_update.run()

        # Source new particles
        imc_source.run()

        # Track new particles through the mesh
        imc_track.run()
        imc_track.clean()

        # Tally
        imc_tally.run()

        # Energy check
        #

        # Update time
        time.time = time.time + time.dt

        # Plot
        if plottimenext <= 2:
            if (time.time * phys.c + 1.0e-06) >= plottimes[plottimenext]:
                print("Plotting {:6d}".format(plottimenext))
                print("at target time {:24.16f}".format(plottimes[plottimenext]))
                print("at actual time {:24.16f}".format(time.time * phys.c))
                # plt.plot(mesh.cellpos, mesh.temp, "b-o")
                fname.write("Time = {:24.16f}\n".format(time.time * phys.c).encode())
                pickle.dump(mesh.cellpos, fname, 0)
                pickle.dump(mesh.temp, fname, 0)
                plottimenext = plottimenext + 1

    # Close file
    fname.close()

    # Plot
    # plt.show()

File: fortran/src/imc_opcon.f90
subroutine imc_opcon

    use imc_global_io_data,   only: output_file
    use imc_global_mesh_data, only: temp0, temp, cellpos
    use imc_global_phys_data, only: c
    use imc_global_time_data, only: time, ns, step, dt

    use imc_utils,            only: imc_get_free_lun

    implicit none

    real(kind=8), dimension(3) :: plottimes = (/ 6.d0, 15.d0, 90.d0 /)
    integer                    :: plottimenext = 1
    integer                    :: fname

    ! Open output file
    fname = imc_get_free_lun()
    open(unit=fname, file=output_file, action='write', form='formatted', status='unknown')

    ! Set an initial temperature field
    temp(:) = temp0(:)  ! keV

    ! Loop over timesteps

    time = 0.d0

    do step = 0, ns - 1

        write(*,'(/,A)')      "==============================================================================="
        write(*,'(A,I4)')     'Timestep  ', step
        write(*,'(A,ES24.16)') 'Time (s)  ', time
        write(*,'(A,ES24.16)') 'Time (cm) ', (time * c)
        write(*,'(A)')        "==============================================================================="

        ! Update temperature-dependent quantities
        call imc_update

        ! Source new particles
        call imc_source

        ! Track new particles through the mesh
        call imc_track

        ! Tally
        call imc_tally

        ! Energy check
        !

        ! Update time
        time = time + dt

        ! Plot
        if (plottimenext <= 3) then
            if ((time * c + 1.d-06) >= plottimes(plottimenext)) then
                write(*,'(A,I6)')     'Plotting ', plottimenext
                write(*,'(A,F24.16)') 'at target time ', plottimes(plottimenext)
                write(*,'(A,F24.16)') 'at actual time ', (time * c)
                write(fname,'(A,F24.16)') 'Time = ',time * c
                write(fname,*) cellpos(:)
                write(fname,*) temp(:)
                plottimenext = plottimenext + 1
            endif
        endif

    enddo

    ! Close output file
    close(fname)

end subroutine imc_opcon
